File: android/widget/src/main/java/com/avatar/inc/widget/AvatarWidgetEvents.kt
package com.avatar.inc.widget

import android.os.Handler
import android.os.Looper
import org.json.JSONObject

class AvatarWidgetEvents {
    var onReady: (() -> Unit)? = null
    var onConversationStarted: ((JSONObject) -> Unit)? = null
    var onMessageReceived: ((JSONObject) -> Unit)? = null
    var onUnreadChanged: ((Int) -> Unit)? = null
    var onError: ((String) -> Unit)? = null

    internal fun dispatch(event: String, payload: JSONObject) {
        Handler(Looper.getMainLooper()).post {
            when (event) {
                "WIDGET_READY" -> onReady?.invoke()
                "CONVERSATION_STARTED" -> onConversationStarted?.invoke(payload)
                "MESSAGE_RECEIVED" -> onMessageReceived?.invoke(payload)
                "UNREAD_CHANGED" -> onUnreadChanged?.invoke(payload.optInt("unreadCount", 0))
                "API_STREAM_ERROR" -> onError?.invoke(payload.optString("reason", "Unknown error"))
            }
        }
    }
}

File: Sources/AvatarWidget/AvatarWidgetEvents.swift
import Foundation

public final class AvatarWidgetEvents {
    public var onReady: (() -> Void)?
    public var onConversationStarted: ((_ payload: [String: Any]) -> Void)?
    public var onMessageReceived: ((_ payload: [String: Any]) -> Void)?
    public var onUnreadChanged: ((_ count: Int) -> Void)?
    public var onError: ((_ reason: String) -> Void)?

    public init() {}

    func dispatch(event: String, payload: [String: Any]) {
        DispatchQueue.main.async {
            switch event {
            case "WIDGET_READY":
                self.onReady?()
            case "CONVERSATION_STARTED":
                self.onConversationStarted?(payload)
            case "MESSAGE_RECEIVED":
                self.onMessageReceived?(payload)
            case "UNREAD_CHANGED":
                let count = payload["unreadCount"] as? Int
                    ?? (payload["unreadCount"] as? NSNumber)?.intValue
                    ?? 0
                self.onUnreadChanged?(count)
            case "API_STREAM_ERROR":
                let reason = payload["reason"] as? String ?? "Unknown error"
                self.onError?(reason)
            default:
                break
            }
        }
    }
}
